### PR TITLE
#174 script to retry failed webhook

### PR DESCRIPTION
**Pull request by GauravGusain98**
https://github.com/DostEducation/RP_IVR_analytics/issues/174

### Types of changes

- [ ] Bug fix (a change which fixes an issue)
- [x] New feature (a change which adds functionality)

### Short description of what this resolves:

- Added functionality to retry failed webhook logs 
- The created on a date is set to the created on the date of logs.

### Checklist:

- [x] I have performed a self-review of my own code.
- [x] The code follows the style guidelines of this project.
- [x] The code changes are passing the CI checks
- [ ] I have documented my code wherever required.
- [ ] The changes requires a change to the documentation.
- [ ] I have updated the documentation based on the my changes.

### CODE REVIEW

File: api/services/call_log_event_service.py [hunk 1]
```diff
@@ -20,6 +21,9 @@ def create_call_log_event(self, data):
             pick_time=data["pick_time"],
             end_time=data["end_time"],
             duration=data["duration"],
+            created_on=data["log_created_on"]
```
Comment (GauravGusain98) [inline api/services/call_log_event_service.py:24]: We will only receive `data["log_created_on"]` for the retry logic.

File: api/services/handle_event_service.py [hunk 2]
```diff
@@ -6,16 +6,15 @@
 
 
 class HandleEventService:
-    def handle_event_service(self, request):
-
+    def handle_event_service(self, form_data, log_created_on=None):
         call_sid_exist = db.session.query(
-            db.exists().where(models.CallLogEvent.call_sid == request.form["CallSid"])
```
Comment (GauravGusain98) [inline api/services/handle_event_service.py:12]: previously we are passing the whole request object but as we have the paylaod as json so to be in sync we are now passing `request.form` for regular webhooks and jsonData for retry webhooks. 

This way both the data can be handle using the same logic

File: api/services/transaction_log_service.py [hunk 3]
```diff
@@ -16,3 +16,12 @@ def mark_ivr_transaction_log_as_processed(self, ivr_transaction_log):
         ivr_transaction_log.processed = True
         db.session.add(ivr_transaction_log)
         db.session.commit()
+
+    def get_failed_ivr_transaction_log(self):
+        failed_ivr_transaction_logs = (
+            models.IvrCallbackTransactionLog.query.filter_by(processed=False)
+            .limit(app.config["RETRY_LOGS_LIMIT"])
```
Comment (GauravGusain98) [inline api/services/transaction_log_service.py:23]: using env for limit

File: config.py [hunk 4]
```diff
@@ -24,7 +24,7 @@
 }
 # For socket based connection
 SQLALCHEMY_DATABASE_URI = (
-    "postgresql://%(user)s:%(password)s/%(database)s?host=%(connection_name)s/"
+    "postgresql://%(user)s:%(password)s@/%(database)s?host=%(connection_name)s/"
```
Comment (GauravGusain98) [inline config.py:27]: Changes related to hotfix. Made another PR for the same.

File: main.py [hunk 5]
```diff
@@ -21,6 +31,30 @@ def callback(request):
...
         return jsonify(message="Something went wrong!"), 400
+
     return "Success"
+
+
+def retry_failed_webhook(transaction_log_service):
+    failed_ivr_logs = transaction_log_service.get_failed_ivr_transaction_log()
+
+    for log in failed_ivr_logs:
+        processed = process_form_data(json.loads(log.payload), log.created_on)
```
Comment (GauravGusain98) [inline main.py:44]: processing each failed log and passing the actual time that be in the created_on column of the new records that will be created.

Comment (Satendra-SR) [reply main.py:44]: Is there any reason to pass created_on separately? Isn't it possible to get it from JSON itself?

File: api/services/transaction_log_service.py (hunk 3 above)
Comment (GauravGusain98) [reply api/services/transaction_log_service.py:23]: 1. Yes In case the logs are too many.
2. 👍

File: main.py (hunk 5 above)
Comment (GauravGusain98) [reply main.py:44]: I don't want to make any changes to the payload that's why I am passing it separately.

File: api/services/transaction_log_service.py [hunk 6]
```diff
@@ -16,3 +16,12 @@ def mark_ivr_transaction_log_as_processed(self, ivr_transaction_log):
         ivr_transaction_log.processed = True
         db.session.add(ivr_transaction_log)
         db.session.commit()
+
+    def get_failed_ivr_transaction_log(self):
+        failed_ivr_transaction_logs = (
```
[the remark this answers is not in the document]
Comment (GauravGusain98) [reply api/services/transaction_log_service.py:21]: > 

payload doesn't make sense as we are getting logs and payload is part of the log

File: main.py [hunk 7]
```diff
@@ -21,6 +31,36 @@ def callback(request):
...
+        payload = json.loads(log.payload)
+        payload["log_created_on"] = log.created_on
+        processed = process_form_data(payload)
+
+        if processed is not True:
+            continue
+
+        log.processed = True
+        db.session.add(log)
+        db.session.commit()
```
Comment (Satendra-SR) [inline main.py:57]: We can reduce the number of database query per execution. Thoughts?
```suggestion
    for log in failed_ivr_logs:
        payload = json.loads(log.payload)
        payload["log_created_on"] = log.created_on
        log.processed = process_form_data(payload)
        log.attempts += 1
        db.session.add(log)
        db.session.commit()
```

**Review by Satendra-SR — APPROVED** [review]
One change is required to reduce the number of DB queries per log processing. Please fix and merge.

File: main.py (hunk 7 above)
Comment (GauravGusain98) [reply main.py:57]: In this way, if the logic is not updating the status?

Comment (Satendra-SR) [reply main.py:57]: Yea, the `log.processed` should be either True or False. And if you see`process_form_data(payload)`, it returns a boolean value. So the status will be based on the return type. Do you think this doesn't work?

**Pull request merged**